Always use Helvetica when rendering PDF forms

ericnormand opened the pull request:
We were having trouble rendering non-ascii characters when filling out
PDF forms. I determined that the standard font used in the form was
MyriadPro, but guess what! It doesn't have special characters, just a
limited set. So when it got rendered, those things got transformed into
some other printable character.

I tried playing with different ways to render them, but settled on this
one. Some of the ways I tried were to upgrade to PDFBox 2.0, which took
a while and I could not get to work. I also considered rendering the
text by hand (instead of using the existing stream).

The solution I settled on, I believe, is the most robust for a variety
of PDFs. First, it adds a new Type 1 Font (Helvetica) called
"Helvetica-DW" to avoid any conflicts with existing fonts. Then it
replaces the font inline into the stream (text replace; PDFs are just
programs in a text format). It replaces any font with the Helvetica-DW
set up earlier.

I'm sure a PDF expert could have done better. This solution has some
consequences: we always render with Helvetica, replacing whatever font
exists; I don't know if this will work if Helvetica is not on the
server. That needs to be tested.

[Pivotal Card](https://www.pivotaltracker.com/story/show/110824994)

ericnormand commented:
A screenshot of the zany characters now working:

![screen shot 2016-01-05 at 2 13 47 pm](https://cloud.githubusercontent.com/assets/116172/12126419/c862accc-b3b6-11e5-8180-3e8d6a7cac14.png)

cap10morgan commented:
Why are the checks in the checkboxes `4` now?

cap10morgan commented on src/form_works/render.clj:
```diff
@@ -68,9 +82,12 @@
+                              (let [s (slurp (.getUnfilteredStream stream))
+                                    s1 (clojure.string/replace s #"/\S+(.*Tf)"
+                                                               (fn [[whole end]]
+                                                                 (str "/Helvetica-DW" end)))]
```
Can this and the code on lines 66-68 be extracted into a common fn?

ericnormand commented:
Fixed the checkmark issue and did some refactoring.

![screen shot 2016-01-06 at 12 02 54 pm](https://cloud.githubusercontent.com/assets/116172/12152367/d4eb530e-b479-11e5-9613-218a2a5ee224.png)

ericnormand commented:
Is the scream because PDFBox is terrible?

cap10morgan commented:
It was referring to the commented-out chunk of code that you have since removed. :)

ericnormand commented:
Let me know what else I can do to get this through.

danielglauser commented:
@ericnormand Did you have to sacrifice a magic buddy to figure this out? Ugh, PDFs.
LGTM

cap10morgan commented:
I'm going over this now.

cap10morgan commented:
Have you tested this on any multi-page documents? [This StackOverflow question](http://stackoverflow.com/questions/25525213/how-to-write-unicode-text-to-pdf-with-pdfbox) and [this PDFBox bug](https://issues.apache.org/jira/browse/PDFBOX-2634) that it links to make me wonder if this will work in PDFBox <2.0 for multi-page docs.

cap10morgan commented:
Are there licensing issues around embedding Helvetica? Are we embedding Helvetica? http://stackoverflow.com/questions/5425251/using-pdfbox-to-write-utf-8-encoded-strings-to-a-pdf/5427096#5427096

cap10morgan commented:
Everything I'm reading says PDFBox 2.0 is much improved on the Unicode front. Since presumably we'll want to update to that major version (or later) at *some* point, I'm curious what issues you ran into getting it to work. Were they Unicode-related, or something that worked before under 1.8.x that you couldn't get working under 2.0?

tie-rack commented:
This doesn't embed Helvetica. Helvetica is one of the standard 14 fonts PDF readers are required to have (or have "acceptable" substitutes for).

ericnormand commented:
@tie-rack is right:

> • PDF prescribes a set of 14 standard fonts that can be used without prior definition.
These include four faces each of three Latin text typefaces (Courier,
Helvetica*, and Times*), as well as two symbolic fonts (Symbol and ITC Zapf
Dingbats®
). These fonts, or suitable substitute fonts with the same metrics, are
required to be available in all PDF consumer applications.

-- http://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf

ericnormand commented:
@cap10morgan 
Re: 2.0: basically the same issues only without anyone having tread the path. As you can see, writing to the PDF is hardly a one-liner, even though this is supposed to be "common practice". This code came from a gist.

2.0 seemed to be improved, and once I converted everything to the new library, I was getting weird a weird issue. Form text widgets had null pages in 2.0. They existed in 1.8. Checkbox widgets had a page in 2.0. I even printed out the raw PDF dictionaries to see what was up, but there indeed was no page in there. Could it be a bug in the parser? I didn't stick around to find out and reverted back to 1.8.

That was one line of attack, and I tried another, which was to change the font. And it turns out that was the real cause: MyriadPro did not have the characters we needed and were being converted into characters it had.

cap10morgan commented:
OK. So the standard Helvetica just happens to have more of the still-Latin-but-not-ASCII characters than whatever font (MyriadPro, I guess) the NVRF form was using before? Will this need to be true on every PDF reader that opens the file for the non-ASCII chars to show up correctly (and is it pretty safe to rely on that)?

Would the same result (for this form only) have been achieved by setting the NVRF PDF we're using to Helvetica instead of MyriadPro using Acrobat or whatever?

ericnormand commented:
@cap10morgan 

Re: multi-page docs: I have not tried it. Do we have a use case? Can that be pushed off into the future? I'd be glad to test it but it could take a couple of days to fix if issues come up.

ericnormand commented:
@cap10morgan 

It needs to be true but is safe to rely on. It's part of the PDF standard.

cap10morgan commented:
@ericnormand I'm not sure how soon we'll need multi-page docs. In Classic they are very often >1 page, but most of the time that's when someone is downloading both a reg. and a VBM form (which are separate downloads for now in the rewrite). So if that's the only use case, we may not need it that soon. But I would like to know very soon if we have a bug lurking there.

ericnormand commented:
I'll look for a multi-page form and make a test for it.

cap10morgan commented:
I discussed this a bit in Slack w/ Eric. This **LGTM** for now. I'm still curious to know if this works w/ multi-page docs or not. But we don't have an immediate use case for those right now AFAIK either.

Hopefully when PDFBox 2.x has been out longer there will be more examples, docs, etc. about making all this work in a less hack-y way (or we'll bite the bullet on iText, but ugh licensing).
